File: Qb.Js/src/From.ts
import { FromTerm } from "./FromTerm";
import { Select } from "./Select";
import { Union } from "./Union";
import { Qb } from "./Qb";
import { InvalidQueryError } from "./InvalidQueryError";


export class From {
  public Term: FromTerm;

  static Table(tableName: string): From;
  static Table(tableName: string, alias: string): From;
  static Table(tableName: string, alias: string, ns: string): From;
  static Table(tableName: string, alias: string, ns1: string, ns2: string): From;
  static Table(): From {
    var tableName: string|undefined;
    var alias: string|undefined;
    var ns1: string|undefined;
    var ns2: string|undefined;

    if (arguments.length >= 1) {
      tableName = arguments[0];
    }
    if (arguments.length >= 2) {
      alias = arguments[1];
    }
    if (arguments.length >= 3) {
      ns1 = arguments[2];
    }
    if (arguments.length >= 4) {
      ns2 = arguments[3];
    }
    var res = new From();
    if(tableName === undefined)
      throw new InvalidQueryError("Table name not specified");
    res.Term = FromTerm.Table(tableName, alias, ns1, ns2);
    return res;
  }

  static SubQuery(subQuery: Select, alias: string): From {
    var res = new From();
    res.Term = FromTerm.SubQuery((<Select>subQuery)["Query"], alias);
    return res;
  }

  static Union(union: Union, alias: string): From {
    var res = new From();
    res.Term = FromTerm.Union(Qb.GetQueryObject(union), alias);
    return res;
  }
}

File: Qb.Net/From.cs
using Viten.QueryBuilder.SqlOm;
using System.Linq;

namespace Viten.QueryBuilder
{
  /// <summary>Класс описания конструкции SELECT ... FROM</summary>
  public class From
  {
    internal FromTerm Term;

    internal From()
    {
    }

    internal From(string tableName)
    {
      Term = FromTerm.Table(tableName);
    }

    internal From(string tableName, string alias)
    {
      Term = FromTerm.Table(tableName, alias);
    }

    internal From(string tableName, string alias, string ns)
    {
      Term = FromTerm.Table(tableName, alias, ns);
    }

    internal From(string tableName, string alias, string ns1, string ns2)
    {
      Term = FromTerm.Table(tableName, alias, ns1, ns2);
    }

    internal From(SelectQuery subQuery, string alias)
    {
      Term = FromTerm.SubQuery(subQuery, alias);
    }

    internal From(OmUnion union, string alias)
    {
      Term = FromTerm.Union(union, alias);
    }

    /// <summary>Описание конструкции SELECT ... FROM</summary>
    public static From Table(string tableName)
    {
      return new From(tableName);
    }

    /// <summary>Описание конструкции SELECT ... FROM</summary>
    public static From Table(string tableName, string alias)
    {
      return new From(tableName, alias);
    }

    /// <summary>Описание конструкции SELECT ... FROM</summary>
    public static From Table(string tableName, string alias, string ns)
    {
      return new From(tableName, alias, ns);
    }

    /// <summary>Описание конструкции SELECT ... FROM</summary>
    public static From Table(string tableName, string alias, string ns1, string ns2)
    {
      return new From(tableName, alias, ns1, ns2);
    }

    /// <summary>Описание конструкции SELECT ... FROM</summary>
    public static From SubQuery(Select subQuery, string alias)
    {
      SelectQuery selectQuery = Qb.GetQueryObject(subQuery);
      return new From(selectQuery, alias);
    }

    /// <summary>Описание конструкции SELECT ... FROM</summary>
    public static From SubQuery(string subQuery, string alias)
    {
      From from = new From();
      from.Term = FromTerm.SubQuery(subQuery, alias);
      return from;
    }

    /// <summary>Описание конструкции SELECT ... FROM</summary>
    public static From Union(Union union, string alias)
    {
      From from = new From();
      from.Term = FromTerm.Union(Qb.GetQueryObject(union), alias);
      return from;
    }

  }
}
